### PR TITLE
Update `sha1`, `sha2`, and `md-5` dependencies

### LINKED ISSUE #93676: Deprecated dependency in tidy's `deps.rs`
According to crates.io, `cpuid-bool` is deprecated. However, it is listed in the permitted dependencies: https://github.com/rust-lang/rust/blob/498eeb72f590e518e19746b346be53713689e207/src/tools/tidy/src/deps.rs#L96

`cpuid-bool` is being pulled in by two hashing crates:

https://github.com/rust-lang/rust/blob/4c55c8362de32be1530b2441c3e3a51e73edeb21/Cargo.lock#L4713-L4737

Both of these crates now use `cpufeatures`.

We should update `sha2` and `sha-1` to use the currently maintained crate.

### PR DESCRIPTION
**Pull request by pierwill**
This replaces the deprecated [`cpuid-bool`](https://crates.io/crates/cpuid-bool) dependency with [`cpufeatures`](https://crates.io/crates/cpufeatures), while adding [`crypto-common`](https://crates.io/crates/crypto-common) as a new dependency.

Closes #93676.

### CODE REVIEW

**Comment by pierwill** [comment]
Just a thought: what about removing `md5`? 🤔 😈

**Comment by klensy** [comment]
But this didn't replaces `cpuid-bool`, as it still presented in Cargo.lock? Plus, this does 2x and 3x for some crate versions, as there still some crates *slow* on updating, like https://github.com/pest-parser/pest/issues/485

**Comment by pierwill** [comment]
> But this didn't replaces cpuid-bool

Fixed 👍

**Comment by klensy** [comment]
As TODO, we can try to see if there any perf changes for for different features, like [this](https://github.com/RustCrypto/hashes/blob/fae1866643425d79569850fd4768015d1bb7a091/sha2/Cargo.toml#L32)

**Comment by klensy** [comment]
Btw, you can bump `md-5` too, as it depends on the same crates.

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -68,9 +68,8 @@ use std::ops::{Add, Range, Sub};
 use std::path::{Path, PathBuf};
 use std::str::FromStr;
 
-use md5::Digest as Md5Digest;
+use md5::Digest;
```
Comment (fbstj) [inline compiler/rustc_span/src/lib.rs:71]: is this used?

File: Cargo.lock [hunk 2]
```diff
@@ -263,7 +272,7 @@ dependencies = [
  "rayon",
  "serde",
  "serde_json",
- "sha2",
+ "sha2 0.9.9",
```
Comment (Mark-Simulacrum) [inline Cargo.lock:275]: Can we bump this as well?

File: compiler/rustc_span/src/lib.rs (hunk 1 above)
Comment (pierwill) [reply compiler/rustc_span/src/lib.rs:71]: `x.py check` fails if this `use` is removed:

```
❯  x.py check
Updating only changed submodules
  Submodules updated in 0.10 seconds
Building rustbuild
    Finished dev [unoptimized] target(s) in 0.42s
Checking stage0 std artifacts (aarch64-apple-darwin -> aarch64-apple-darwin)
    Finished release [optimized] target(s) in 0.25s
Checking stage0 std test/bench/example targets (aarch64-apple-darwin -> aarch64-apple-darwin)
    Finished release [optimized] target(s) in 0.25s
Checking stage0 compiler artifacts (aarch64-apple-darwin -> aarch64-apple-darwin)
    Checking rustc_span v0.0.0 (/Users/repos/rust/compiler/rustc_span)
error[E0599]: no function or associated item named `digest` found for struct `CoreWrapper` in the current scope
    --> compiler/rustc_span/src/lib.rs:1301:45
     |
1301 |                 value.copy_from_slice(&Md5::digest(data));
     |                                             ^^^^^^ function or associated item not found in `CoreWrapper<Md5Core>`
     |
     = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
     |
32   | use md5::Digest;
     |

error[E0599]: no function or associated item named `digest` found for struct `CoreWrapper` in the current scope
    --> compiler/rustc_span/src/lib.rs:1304:46
     |
1304 |                 value.copy_from_slice(&Sha1::digest(data));
     |                                              ^^^^^^ function or associated item not found in `CoreWrapper<Sha1Core>`
     |
     = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
     |
32   | use md5::Digest;
     |

error[E0599]: no function or associated item named `digest` found for struct `CoreWrapper` in the current scope
    --> compiler/rustc_span/src/lib.rs:1307:48
     |
1307 |                 value.copy_from_slice(&Sha256::digest(data));
     |                                                ^^^^^^ function or associated item not found in `CoreWrapper<CtVariableCoreWrapper<Sha256VarCore, UInt<UInt<UInt<UInt<UInt<UInt<UTerm, B1>, B0>, B0>, B0>, B0>, B0>>>`
     |
     = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
     |
32   | use md5::Digest;
     |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustc_span` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:00:07
```

**Pull request merged**